### PR TITLE
Added favorites star in owner field (select2)

### PR DESCRIPTION
**Pull request by adudekyeti**
Added favorites star in owner field (select2)

### CODE REVIEW

File: public_html/layouts/resources/Fields.js [hunk 1]
```diff
@@ -697,6 +707,50 @@ App.Fields = {
...
+					currentOptionId = '';
+				if (idArr.length > 0) {
+					currentOptionId = idArr[idArr.length - 1];
+				}
+				let params = {
+					userId: currentOptionId
+				};
+				//AppConnector.request(params).done(function (data) {
+				if (thisInstance.data('prefix') === 'far') {
+					thisInstance.addClass('fas').removeClass('far');
```
Comment (rskrzypczak) [inline public_html/layouts/resources/Fields.js]: triggerClass

File: public_html/layouts/resources/Fields.js [hunk 2]
```diff
@@ -697,6 +707,50 @@ App.Fields = {
...
+			element.on('mouseup', '[data-fa-i2svg].js-user-favorites', function (event) {
+				let thisInstance = $(this),
+					liElement = thisInstance.closest('.select2-results__option'),
+					id = liElement.attr("id"),
+					idArr = id.split("-"),
+					currentOptionId = '';
+				if (idArr.length > 0) {
+					currentOptionId = idArr[idArr.length - 1];
+				}
+				let params = {
```
Comment (rskrzypczak) [inline public_html/layouts/resources/Fields.js]: params = ... .data('url')

File: public_html/layouts/resources/Fields.js [hunk 3]
```diff
@@ -697,6 +707,50 @@ App.Fields = {
...
+				if (liElement.hasClass('group-result')) {
+					return;
+				}
+				if (event.type === 'mouseenter' && liUserOptions.length === 0) {
+					this.appendOptionActionsTemplate(liElement);
+				}
+			});
+		},
+		userOptionClickEvent(element) {
+			element.on('mouseup', '[data-fa-i2svg].js-user-favorites', function (event) {
```
Comment (rskrzypczak) [inline public_html/layouts/resources/Fields.js]: general class name (js-user-favorites => js-select-option-event ?)

File: public_html/layouts/resources/Fields.js [hunk 4]
```diff
@@ -697,6 +707,50 @@ App.Fields = {
 				}
 			});
 		},
+		registerCustomFilterOptionsHoverEvent(element) {
+			element.on('mouseenter mouseleave', 'li.select2-results__option[role="treeitem"]', (event) => {
+				let liElement = $(event.currentTarget),
+					liUserOptions = liElement.find('.js-user-actions');
```
Comment (rskrzypczak) [inline public_html/layouts/resources/Fields.js]: change class name 'js-user-actions'

File: public_html/layouts/resources/Fields.js [hunk 5]
```diff
@@ -697,6 +707,50 @@ App.Fields = {
...
+				if (liElement.hasClass('group-result')) {
+					return;
+				}
+				if (event.type === 'mouseenter' && liUserOptions.length === 0) {
+					this.appendOptionActionsTemplate(liElement);
+				}
+			});
+		},
+		userOptionClickEvent(element) {
```
Comment (rskrzypczak) [inline public_html/layouts/resources/Fields.js]: change function name

File: public_html/layouts/resources/Fields.js [hunk 6]
```diff
@@ -697,6 +707,50 @@ App.Fields = {
...
+				} else {
+					thisInstance.addClass('far').removeClass('fas');
+					thisInstance.data('prefix', 'far');
+				}
+				//});
+				event.stopPropagation();
+			});
+		},
+		appendOptionActionsTemplate(liElement) {
+			let template = $(`<span class="js-user-actions o-filter-actions noWrap float-right">
```
Comment (rskrzypczak) [inline public_html/layouts/resources/Fields.js]: change js-user-actions, data-value="favorites", js-user-favorites to more general.
get icons from .data() elements

File: layouts/basic/modules/Vtiger/Edit/Field/Owner.tpl [hunk 7]
```diff
@@ -44,7 +50,10 @@
 					data-name="{$ASSIGNED_USER_ID}" name="{$ASSIGNED_USER_ID}" data-fieldinfo='{$FIELD_INFO}'
 					{if !empty($SPECIAL_VALIDATOR)}data-validator={\App\Json::encode($SPECIAL_VALIDATOR)}{/if} {if $FIELD_MODEL->isEditableReadOnly()}readonly="readonly"{/if} {if $USER_MODEL->isAdminUser() == false && $ROLE_RECORD_MODEL->get('changeowner') == 0}readonly="readonly"{/if}
 					{if AppConfig::performance('SEARCH_OWNERS_BY_AJAX')}
-				data-ajax-search="1" data-ajax-url="index.php?module={$MODULE}&action=Fields&mode=getOwners&fieldName={$ASSIGNED_USER_ID}" data-minimum-input="{AppConfig::performance('OWNER_MINIMUM_INPUT_LENGTH')}"
+						data-ajax-search="1" data-ajax-url="index.php?module={$MODULE}&action=Fields&mode=getOwners&fieldName={$ASSIGNED_USER_ID}" data-minimum-input="{AppConfig::performance('OWNER_MINIMUM_INPUT_LENGTH')}"
+					{/if}
+					{if AppConfig::module('Users','FAVORITE_OWNERS')}
```
Comment (rskrzypczak) [inline layouts/basic/modules/Vtiger/Edit/Field/Owner.tpl]: elseif

File: layouts/basic/modules/Vtiger/Edit/Field/Owner.tpl [hunk 8]
```diff
@@ -30,7 +30,13 @@
 					{foreach key=OWNER_ID item=OWNER_NAME from=$OWNERS}
 						<option value="{$OWNER_ID}"
 								data-picklistvalue="{$OWNER_NAME}" {if $FIELD_VALUE eq $OWNER_ID} selected {/if}
-								data-userId="{$CURRENT_USER_ID}">
+								data-userId="{$CURRENT_USER_ID}"
+								{if AppConfig::module('Users','FAVORITE_OWNERS')}
```
Comment (rskrzypczak) [inline layouts/basic/modules/Vtiger/Edit/Field/Owner.tpl]: go to assign

File: public_html/layouts/resources/Fields.js [hunk 9]
```diff
@@ -533,6 +533,10 @@ App.Fields = {
 					if (data.element && data.element.className) {
 						$(container).addClass(data.element.className);
 					}
+					let actualElement = $(data.element);
+					if (typeof selectElement.data('showAdditionalIcons') !== "undefined" && actualElement.is('option')) {
+						return '<div class="js-user__title d-flex justify-content-between" data-js="appendTo"><div class="u-text-ellipsis--no-hover">' + actualElement.text() + '</div></div>';
```
Comment (rskrzypczak) [inline public_html/layouts/resources/Fields.js]: js-user__title

File: public_html/layouts/resources/Fields.js [hunk 10]
```diff
@@ -697,6 +707,54 @@ App.Fields = {
 				}
 			});
 		},
+		registerCustomFilterOptionsHoverEvent(selectElement) {
```
Comment (rskrzypczak) [inline public_html/layouts/resources/Fields.js]: short

File: public_html/layouts/resources/Fields.js [hunk 11]
```diff
@@ -697,6 +707,54 @@ App.Fields = {
...
+			const self = this;
+			let select2Instance = selectElement.data('select2');
+			select2Instance.$dropdown.on('mouseenter mouseleave', 'li.select2-results__option[role="treeitem"]', (event) => {
+				let liElement = $(event.currentTarget),
+					liSelectOptions = liElement.find('.js-select-option-actions');
+				if (liElement.hasClass('group-result')) {
+					return;
+				}
+				if (event.type === 'mouseenter' && liSelectOptions.length === 0) {
```
Comment (rskrzypczak) [inline public_html/layouts/resources/Fields.js]: !liElement.hasClass('group-result')

File: public_html/layouts/resources/Fields.js [hunk 12]
```diff
@@ -697,6 +707,54 @@ App.Fields = {
...
+		},
+		appendOptionClickEvent(optionElement, element) {
+			element.on('mouseup', '[data-fa-i2svg].js-select-option-event', function (event) {
+				let thisInstance = $(event.currentTarget),
+					params = optionElement.data('url'),
+					iconActive = optionElement.data('iconActive'),
+					iconInactive = optionElement.data('iconInactive');
+				//AppConnector.request(params).done(function (data) {
+				if (optionElement.data('state') === 'active') {
+					thisInstance.toggleClass(iconActive).toggleClass(iconInactive);
```
Comment (rskrzypczak) [inline public_html/layouts/resources/Fields.js]: thisInstance.toggleClass(iconActive+' '+iconInactive)

File: public_html/layouts/resources/Fields.js [hunk 13]
```diff
@@ -697,6 +707,54 @@ App.Fields = {
...
+				event.stopPropagation();
+			});
+		},
+		appendOptionActionsTemplate(optionElement, liElement) {
+			let optionClass = optionElement.data('state') === 'active' ? optionElement.data('icon-active') : optionElement.data('icon-inactive'),
+				template = $(`<span class="js-select-option-actions o-filter-actions noWrap float-right">
+					<span data-js="click | class: fas | far" class="mr-1 js-select-option-event ${optionClass}"></span>
+				</span>`);
+			this.appendOptionClickEvent(optionElement, template);
+			template.appendTo(liElement.find('.js-user__title'));
```
Comment (rskrzypczak) [inline public_html/layouts/resources/Fields.js]: js-user__title

File: public_html/layouts/resources/Fields.js [hunk 14]
```diff
@@ -697,6 +707,51 @@ App.Fields = {
...
 		},
+		registerElementOptions(selectElement) {
+			const self = this;
+			let select2Instance = selectElement.data('select2');
+			select2Instance.$dropdown.on('mouseenter mouseleave', 'li.select2-results__option[role="treeitem"]', (event) => {
+				let liElement = $(event.currentTarget),
+					liSelectOptions = liElement.find('.js-select-option-actions');
+				if (!liElement.hasClass('group-result') && event.type === 'mouseenter' && liSelectOptions.length === 0) {
+					let select2Option = liElement.closest('.select2-results__option'),
+						id = select2Option.attr("id"),
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/resources/Fields.js]:
```suggestion
						id = select2Option.attr('id'),
```

File: public_html/layouts/resources/Fields.js [hunk 15]
```diff
@@ -697,6 +707,51 @@ App.Fields = {
...
+						idArr = id.split("-"),
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/resources/Fields.js]: ?

File: public_html/layouts/resources/Fields.js [hunk 16]
```diff
@@ -697,6 +707,51 @@ App.Fields = {
...
+					iconInactive = optionElement.data('iconInactive');
+				//AppConnector.request(params).done(function (data) {
+				if (optionElement.data('state') === 'active') {
+					thisInstance.toggleClass(iconActive + ' ' + iconInactive);
+					optionElement.data('state', 'inactive');
+				} else {
+					thisInstance.toggleClass(iconInactive + ' ' + iconActive);
+					optionElement.data('state', 'active');
+				}
+				//});
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/resources/Fields.js]: ?

File: public_html/layouts/resources/Fields.js [hunk 17]
```diff
@@ -697,6 +707,51 @@ App.Fields = {
...
+				}
+			});
+		},
+		appendOptionClickEvent(optionElement, element) {
+			element.on('mouseup', '.js-select-option-event', function (event) {
+				let thisInstance = $(event.currentTarget),
+					params = optionElement.data('url'),
+					iconActive = optionElement.data('iconActive'),
+					iconInactive = optionElement.data('iconInactive');
+				//AppConnector.request(params).done(function (data) {
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/resources/Fields.js]: ?

**Pull request merged**